workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/SpatialOSBehaviourLibrary.cs

```diff
@@ -0,0 +1,166 @@
+﻿using System;
+using System.Collections.Generic;
+using Unity.Entities;
+using UnityEngine;
+
+namespace Improbable.Gdk.Core
+{
+    /// <summary>
+    /// Retrieves Reader and Writer fields from MonoBehaviours and handles injection into them.
+    /// </summary>
+    public class SpatialOSBehaviourLibrary {
+        // Stores maps from Spatial component IDs to MemberAdapters to reflectively set Readers/Writers for each SpatialOSBehaviour.
+        private readonly Dictionary<Type, Dictionary<uint, IMemberAdapter>> adapterCache
+            = new Dictionary<Type, Dictionary<uint, IMemberAdapter>>();
+
+        // List of component IDs for which a reader/writer is required for each SpatialOSBehaviour
+        private readonly Dictionary<Type, List<uint>> componentReaderIdsForBehaviours = new Dictionary<Type, List<uint>>();
+        private readonly Dictionary<Type, List<uint>> componentWriterIdsForBehaviours = new Dictionary<Type, List<uint>>();
+
+        // Helper objects for querying members with the [Require] attribute
+        private readonly MemberReflectionUtil reflectionUtil = new MemberReflectionUtil(typeof(RequireAttribute));
+
+        private readonly HashSet<Type> badTypes = new HashSet<Type>();
+
+        private readonly ILogDispatcher logger;
+        private const string LoggerName = "SpatialOSBehaviourLibrary";
+
+        private const string BadRequiredMemberWarning
+            = "[Require] attribute found on member that is not Reader or Writer or incorrectly generated, ignoring!";
+
+        private const string MultipleReadersWritersRequiredError
+            = "MonoBehaviour found requesting more than one view (Reader or Writer) to the same component, " +
+            "this is invalid and will not be enabled!";
+
+        public SpatialOSBehaviourLibrary(ILogDispatcher logger)
+        {
+            this.logger = logger;
+        }
+
+        public void InjectAllReadersWriters(MonoBehaviour spatialOSBehaviour)
```

Comment:
Where would this be called?

Reply:
By the GameObjectRepresentation ECS component, which will keep track of which MonoBehaviours on the corresponding GameObjects require which readers/writers and when those are available.

Reply:
By methods on the GameObjectRepresentation ECS component, which will be keeping track of which readers/writers are needed by the behaviours on its gameobject, and when those are all available (components exist and authority for writers)

Reply:
See also the 3rd design doc for readers/writers, this is sketched there pretty accurately

Reply:
I don't think the actual code in that doc was the thing that was signed off. It was the issues raised. But more importantly, where would you inject all the readers and writers? Wouldn't you inject only the ones that become available?

Reply:
As far as I know the plan is to keep track of which / how many are still not ready, and when everything is ready only then inject everything and enable the monobehaviour. I.e. when a new component is added or authority is received and the GameObjectReference sees that this was the last remaining piece, it calls this function.

Reply:
Required readers/writers in a monobehaviour shall be either all injected or all null at all times as signed off in the design doc.

Reply:
I don't think that is what we discussed in the meetings. That reintroduces old race conditions.

Reply:
This was 100% discussed. Do we really need to start recording every meeting now? What are the race conditions you talked about? Also, these issues could have been pointed out way earlier. We had the design doc and the production meeting where this was mentioned.

Reply:
The change required to this code is really minimal though, we'd literally just need to make the private Inject()/DeInject() methods public, and could remove the InjectAll() / DeInjectAll() methods if they are not used. I'm not sure about the discussion but I don't think it should block this PR anyway.

Reply:
was there a conclusion to that?